### PR TITLE
Validate volume properties before creating VD

### PR DESCRIPTION
**Pull request by Rachit7194**
Hi Chris,

I have added a method which validates volume's size for each of the volumes before creating virtual disks.

Please have a look.

-Rachit

### CODE REVIEW

**Comment by Rachit7194** [comment]
Hi Chris, 

I have execute this change with controller and compute nodes it executing successfully but failing into storage nodes. i will debug that more.

I also have one doubt which is not getting clear in Jetpack/assign_role.py at line :[#L461](https://github.com/dsp-jetpack/JetPack/blob/experimental_upstream_ironic/src/pilot/assign_role.py#L461)
i.e when logical disk gets created for boss controller, it assigns it's size by 0

Thanks,
Rachit

**Comment by cdearborn** [comment]
Hey Rachit,

specifying a size of 0 for the BOSS card when creating the virtual disk is a quirk of the BOSS card.  If you specify a size of anything other than 0, then creation of the virtual disk will fail.

**Comment by Rachit7194** [comment]
> Hey Rachit,
> 
> specifying a size of 0 for the BOSS card when creating the virtual disk is a quirk of the BOSS card. If you specify a size of anything other than 0, then creation of the virtual disk will fail.

Okay ...Thanks Chris

**Comment by Rachit7194** [comment]
**### Just adding the explanation for purpose of code change:** 
 
-----------------------------------CREATE CONFIGURATION-------------------------------------------------
OVERVIEW : 

1. Now create_config clean_step will execute two configuration job till it create VD:
     - Job for drives conversion (if controller supports)
	 - Job for VD creation
2. However, we have a hidden bug in idrac firmware where size of the physical disks increases by some MBs
   from its actual sizes, only when we convert all disks from RAID to JBOD mode.
   Also it comes back to it's actual size when physical drives are converted from JBOD to RAID
3. In create_configuration, 
		1. find appropriate physical_disks and logical_disks and its size using  _find_configuration()
		2. convert_disks_to_raid()
		3. create_virtual_disk()

####################### ISSUE FOUND AND INVESTIGATED WHEN create_configuration clean_step called #########


1. Let's assume we have physical disks converted JBOD mode which means idrac's bug acquired on all physical disks which are
   converted to JBOD.

2. As per point 3.1 above,  _find_configuration() calculates physical disks and logical_disk and it's size with
  respect to physical disks attached in logical_disks.

3. Now the volume size calculated for virtual disks will be according to physical disks(disks having extra sizes)

4. Thenafter, drives conversion get called and all physical disks are converted back to RAID from JBOD 
   with its actual respective sizes
   
--------HERE COMES THE TURNING POINT-------------------------------------------

5. (IN POINT 3) But the volume size calculated to create VD on respective physical disks will be more than 
   actual size of the physical disks on which VD is to be created.

6. Hence Virtual DISK creation fails.

################################ ISSUE RESOLUTION #########################################

1. Make a validation call after drives conversion which validation size of the each logical disk dictionary.
   validate_volume_size()
   
--------------- THINGS WE HAVE ------------------------------------------

2. We will also have logical_disk with its calculated size(which is wrong) on the basis of above physical_disks.

---------------------------------------------------------------------------------------------------------------

------------ THINGS TO DO-----------------------------------------------------------------------------------
3. In validation call (validate_volume_size()), iterate each logical disk and take out it's size(size_mb).

4. Create a new object of physical_disks after drives converted back to RAID so that we can get the
   actual size of physical disks.

5. Filter out the new physical disks attach to each logical disks respectively.

6. calculate the the size required to create logical disks by calling _max_volume_size_mb()
   i.e., new_size_calculated = _max_volume_size_mb()

7. Compare the  new size calculated(above in point 7) with size_mb(size) of each logical disk 

	8. If logical_disk['size_mb'] is different then the new size of logical(calculated in point 7)
	   then call _calculate_volume_props() again to calculate its volume properties by attaching
	   the new size calculated to each logical disk(i.e logical_disk['size_mb'] = new_size_calculated
	
	8.1) append each _new_logical disk to the list of new_logical_disks 

10. In the end if point 9 satisfies and new_logical_disks list is not None, return new logical_disks
    from the validate method i.e; validate_volume_size()  to create Virtual disks.
	
11. If point 9 does not satisfies return the logical_disks parameter which (validate_volume_size())
    it takes as an argument, as is.

**Comment by Rachit7194** [comment]
This PR is dependent on #10 because i have modification on this PR

File: ironic/drivers/modules/drac/raid.py [hunk 1]
```diff
@@ -1064,8 +1103,12 @@ def create_configuration(self, task,
                 "logical disks to RAID mode for node %(node_uuid)s",
                 {'node_uuid': node.uuid})
 
+        #Check valid properties attached to volumes
+        if conversion_results:
+            logical_disks = _validate_volume_size(node, logical_disks_to_create)
+
         controllers = list()
-        for logical_disk in logical_disks_to_create:
+        for logical_disk in logical_disks:
```
Comment (cdearborn) [inline ironic/drivers/modules/drac/raid.py]: I don't think the change on the line above is correct.  If _validate_volume_size() is not called, then this change will cause it to create all of the virtual disks in logical_disks, which is incorrect.  I think the change on the line above should be reverted, and line 1108 should be changed to:
logical_disks_to_create = _validate...

**Review by cdearborn — COMMENTED** [review]
Can you change the permissions on this file back to 0644?
Whenever you log something, be sure to always include some identifying information in the log message, not just a fixed text string.  Ironic is very multithreaded, so imagine looking in the log and seeing a fixed text string - it's not very useful if you can't identify what node it's against, etc.
I think this solution is ok, but one thing that could happen which would be unexpected is that if the user asked for a RAID volume of a specific size, the _validate_volume_size() method could end up making the created RAID volume be a smaller size than what the user requested.  If the user just asked for MAX, then the code will work fine.  It's when they ask for a specific size that there could be an issue.
I'm wondering if there is a way to calculate what the size of the physical disk would be in RAID mode if it's currently in JBOD mode.  If so, then a much simpler solution would be to change the code in _find_configuration() so that it uses the size of the disk as report if it's in RAID mode, and if it's in JBOD mode, then calculate what the size of the disk will be after conversion to RAID mode and use that size.
This would make it so we wouldn't have to change any of the following code and it would eliminate the issue where the user could ask for a volume of one size and end up with a volume with a slightly smaller size.  I'll ask the RAID team if there is a simple way to calculate the disk size and see if this approach is feasible.

**Comment by Rachit7194** [comment]
Thanks Chris for the review. I will take care of adding log statements very precisely next time.

File: ironic/drivers/modules/drac/raid.py (hunk 1 above)
Comment (Rachit7194) [reply ironic/drivers/modules/drac/raid.py]: done

**Comment by Rachit7194** [comment]
Hi Chris,
As you said above in last sentence :

> I'll ask the RAID team if there is a simple way to calculate the disk size and see if this approach is feasible.

Do we have any other solution to calculate disk size or should we considered this solution as good to go?

Thanks,
Rachit

**Comment by cdearborn** [comment]
Hey Rachit,

I have been going back and forth with the RAID team, and have found out:
For the PERC H730P:
If a disk is in RAID/Ready mode:
- SizeInBytes and FreeSizeInBytes are the same, and are the size of the disk in RAID mode

If a disk is in non-RAID/JBOD mode:
- SizeInBytes is the raw size of the disk (this is the size that's too big)
- FreeSizeInBytes is the size of the disk if you convert it to RAID mode

For the PERC H740P:
In RAID mode:
- SizeInBytes is the size of the disk in RAID mode
- FreeSizeInBytes is 0

In eHBA mode:
- SizeInBytes is the size of the disk in RAID mode
- FreeSizeInBytes "may be 0 until you convert it to RAID"

So, for the H730P, we should always use FreeSizeInBytes and for the H740P, we should always use SizeInBytes.  UGH!!!  If we did that, then I think it would fix the problem.  That type of code should be buried in python-dracclient thought, so the ironic code doesn't have to have knowledge of the RAID controller model.  Still digesting this at this point, so not sure what approach we should take.  Your ideas are welcome.

**Review by cdearborn — APPROVED** [review]
Hey Rachit, I spent quite a bit of time finding out how we could get the size of the disk in RAID mode when it's in JBOD mode, and then spent quite a bit of time testing.  In the end, I'm not confident that the values reported back are correct.  As a result, I think we should go with the fix that you have done here.  Please go ahead and submit this patch upstream.

Nice work by the way!

**Comment by Rachit7194** [comment]
Thanks Chris, I will start working on unit testcases and will raise an upstream patch.

**Comment by cdearborn** [comment]
Closing this PR since it has been submitted upstream:
https://review.opendev.org/710852

**Pull request closed without merging**